### PR TITLE
More closely follow hb-repacker when isolating/assigning spaces

### PR DESCRIPTION
**Pull request by cmyr**
My previous implementation was a looser interpretation of hb, but for the promotion/splitting work I want to have as much of the general code match HB has possible, so I can know bugs are in the implementation, not in the design.

based on #385, which should go in first.

### CODE REVIEW

File: write-fonts/src/graph.rs [hunk 1]
```diff
@@ -408,157 +409,208 @@ impl Graph {
...
 
-        // assign all nodes reachable from 16/24 bit edges to space 0.
-        self.assign_space_0();
+        log::debug!("found {} space roots to isolate", roots.len());
 
-        while !roots.is_empty() {
-            let root = *roots.iter().next().unwrap();
-            self.isolate_and_assign_space(root);
-            roots.remove(&root);
+        // we want to *invert* the visited set, but we don't have a fancy hb_set_t
```
Comment (rsheeter) [inline write-fonts/src/graph.rs:430]: should we file an issue to write an hb_set_t? - I believe some of hb's structures are very carefully tuned to work well for text.

File: write-fonts/src/graph.rs [hunk 2]
```diff
@@ -408,157 +409,208 @@ impl Graph {
...
+        let _all_ids = self.order.iter().copied().collect::<HashSet<_>>();
```
Comment (rsheeter) [inline write-fonts/src/graph.rs:431]: why did you need to bind _all_ids? It looks like you only use it the only time?

**Review by rsheeter — APPROVED** [review]
I won't claim full understanding of the algorithms - if you want that Garret needs to review - but the Rust code seems reasonable enough.

LGTM with the same request as the prior CL: lets be very dilligent about linking to the prior art on a fine grained basis. https://github.com/googlefonts/fontmake-rs/blob/main/fontir/src/variations.rs shows my attempt to do this for some of the py:rust conversion.

File: write-fonts/src/graph.rs (hunk 1 above)
Comment (cmyr) [reply write-fonts/src/graph.rs:430]: We've got one (#192)
